File: src/codex_team/daemon/health.py
from __future__ import annotations

import asyncio
import time

from codex_team.config import Config
from codex_team.daemon.event_bus import EventBus
from codex_team.daemon.registry import RegistryStore
from codex_team.schemas.registry import SessionStatus


class HealthMonitor:
    def __init__(
        self,
        *,
        cfg: Config,
        registry: RegistryStore,
        sessions: dict,
        event_bus: EventBus,
        factory,
    ) -> None:
        self._cfg = cfg
        self._registry = registry
        self._sessions = sessions
        self._event_bus = event_bus
        self._factory = factory
        self._healed_at: dict[str, float] = {}

    async def tick_once(self) -> None:
        semaphore = asyncio.Semaphore(max(1, self._cfg.heartbeat.health_check_concurrency))

        async def check_entry(entry) -> None:
            if entry.status == SessionStatus.closed:
                return
            session = self._sessions.get(entry.name)
            if session is None:
                await self._on_down(entry.name)
                return
            try:
                async with semaphore:
                    if not session.is_transport_alive():
                        raise RuntimeError("transport is not alive")
                    await asyncio.wait_for(
                        session.health_check(),
                        timeout=self._cfg.heartbeat.health_timeout_seconds,
                    )
            except Exception as exc:  # noqa: BLE001
                self._registry.update(entry.name, status=SessionStatus.errored, error_message=str(exc))
                await self._on_down(entry.name, session=session)

        await asyncio.gather(*(check_entry(entry) for entry in self._registry.list()))

    async def _on_down(self, name: str, *, session=None) -> None:
        entry = self._registry.get(name)
        last_healed_at = self._healed_at.get(name)
        can_attempt_heal = (
            self._cfg.heartbeat.self_heal_once
            and self._factory is not None
            and (
                last_healed_at is None
                or time.monotonic() - last_healed_at >= self._cfg.heartbeat.self_heal_backoff_seconds
            )
        )
        if can_attempt_heal:
            self._healed_at[name] = time.monotonic()
            try:
                session = await asyncio.wait_for(
                    self._factory.resume(name),
                    timeout=self._cfg.heartbeat.resume_timeout_seconds,
                )
            except Exception as exc:  # noqa: BLE001
                self._registry.update(name, status=SessionStatus.errored, error_message=str(exc))
            else:
                self._sessions[name] = session
                self._registry.update(name, status=SessionStatus.idle, error_message=None)
                self._event_bus.publish("events", {"kind": "auto-heal", "session": name})
                return
        self._event_bus.publish(
            "events",
            {
                "kind": "session-down",
                "session": name,
                "last_error": entry.error_message or "",
                "stderr_tail": session.stderr_tail(limit=20) if session is not None else "",
            },
        )

File: plugins/codex-team/src/health.ts
import { Config } from "./config";
import { EventBus } from "./eventBus";
import { RegistryStore } from "./registry";
import { Session } from "./session";
import { DEFAULT_WORKSPACE, workspaceSessionKey } from "./workspace";

interface SessionFactoryLike {
  resume(name: string, workspace?: string): Promise<Session>;
}

export class HealthMonitor {
  private readonly healedAt = new Map<string, number>();
  private readonly stuckTurnNotified = new Map<string, string>();

  constructor(
    private cfg: Config,
    private readonly registry: RegistryStore,
    private readonly sessions: Map<string, Session>,
    private readonly eventBus: EventBus,
    private readonly factory: SessionFactoryLike,
  ) {}

  replaceConfig(cfg: Config): void {
    this.cfg = cfg;
  }

  async tickOnce(): Promise<void> {
    const entries = this.registry.list(null, true);
    const concurrency = Math.max(1, this.cfg.heartbeat.healthCheckConcurrency);
    let index = 0;
    const workers = Array.from({ length: Math.min(concurrency, entries.length || 1) }, async () => {
      while (index < entries.length) {
        const current = entries[index];
        index += 1;
        await this.checkEntry(current);
      }
    });
    await Promise.all(workers);
  }

  private async checkEntry(entry: ReturnType<RegistryStore["get"]>): Promise<void> {
    const key = sessionKey(entry.workspace, entry.name);
    if (entry.status === "closed") {
      this.stuckTurnNotified.delete(key);
      return;
    }
    const session = this.sessions.get(key) || this.sessions.get(entry.name);
    if (!session) {
      this.stuckTurnNotified.delete(key);
      await this.onDown(entry.workspace, entry.name);
      return;
    }

    this.maybeEmitTurnStuck(entry.workspace, entry.name, session);

    try {
      if (!session.isTransportAlive()) {
        throw new Error("transport is not alive");
      }
      await withTimeout(session.healthCheck(), this.cfg.heartbeat.healthTimeoutSeconds * 1000);
    } catch (error) {
      this.registry.update(entry.name, {
        status: "errored",
        errorMessage: (error as Error).message,
      }, entry.workspace);
      await this.onDown(entry.workspace, entry.name, session);
    }
  }

  private maybeEmitTurnStuck(workspace: string, name: string, session: Session): void {
    const key = sessionKey(workspace, name);
    if (!session.isRunning()) {
      this.stuckTurnNotified.delete(key);
      return;
    }
    const turnId = session.currentTurnId();
    const ageMs = session.currentTurnAgeMs();
    if (!turnId || ageMs == null) {
      this.stuckTurnNotified.delete(key);
      return;
    }
    const thresholdMs = this.cfg.heartbeat.turnStuckSeconds * 1000;
    if (ageMs < thresholdMs) {
      if (this.stuckTurnNotified.get(key) !== turnId) {
        this.stuckTurnNotified.delete(key);
      }
      return;
    }
    if (this.stuckTurnNotified.get(key) === turnId) {
      return;
    }
    this.stuckTurnNotified.set(key, turnId);
    this.eventBus.publish("events", {
      workspace,
      kind: "turn-stuck",
      session: name,
      turn_id: turnId,
      age_ms: ageMs,
      threshold_ms: thresholdMs,
    });
  }

  private async onDown(workspace: string, name: string, session?: Session): Promise<void> {
    const key = sessionKey(workspace, name);
    const entry = this.registry.get(name, workspace);
    const lastHealedAt = this.healedAt.get(key);
    const duringTurn = session?.isRunning() || entry.status === "running";
    const activeTurnId = session?.currentTurnId() || entry.lastTurnId || null;
    const turnAgeMs = session?.currentTurnAgeMs() ?? null;
    const migratedQueue = session ? await session.detachForRecovery("auto-heal queue migration") : [];
    if (session) {
      this.sessions.delete(key);
      this.sessions.delete(name);
    }
    const canAttemptHeal =
      this.cfg.heartbeat.selfHealOnce &&
      !entry.ephemeral &&
      (lastHealedAt == null ||
        Date.now() - lastHealedAt >= this.cfg.heartbeat.selfHealBackoffSeconds * 1000);
    if (canAttemptHeal) {
      this.healedAt.set(key, Date.now());
      try {
        const resumed = await withTimeout(
          this.factory.resume(name, workspace),
          this.cfg.heartbeat.resumeTimeoutSeconds * 1000,
        );
        await resumed.absorbQueue(migratedQueue);
        this.sessions.set(key, resumed);
        this.registry.update(name, {
          status: "idle",
          errorMessage: null,
        }, workspace);
        this.eventBus.publish("events", {
          workspace,
          kind: duringTurn ? "auto-heal-after-crash" : "subprocess-recycled",
          session: name,
          heal_reason: duringTurn ? "transport_down_during_turn" : "transport_down_idle",
          was_during_turn: duringTurn,
          turn_id: activeTurnId,
          turn_age_ms: turnAgeMs,
        });
        return;
      } catch (error) {
        for (const item of migratedQueue) {
          item.waitRejecter?.(new Error(`auto-heal failed for ${name}: ${(error as Error).message}`));
        }
        this.registry.update(name, {
          status: "errored",
          errorMessage: (error as Error).message,
        }, workspace);
      }
    }
    if (!canAttemptHeal) {
      for (const item of migratedQueue) {
        item.waitRejecter?.(new Error(`session ${name} went down and could not be auto-healed`));
      }
    }
    this.eventBus.publish("events", {
      workspace,
      kind: "session-down",
      session: name,
      reason: duringTurn ? "transport_down_during_turn" : "transport_down_idle",
      was_during_turn: duringTurn,
      turn_id: activeTurnId,
      turn_age_ms: turnAgeMs,
      queued_items_migrated: migratedQueue.length,
      lastError: entry.errorMessage || "",
      stderrTail: session?.stderrTail(20) || "",
    });
  }
}

function sessionKey(workspace: string | undefined, name: string): string {
  return workspaceSessionKey(workspace || DEFAULT_WORKSPACE, name);
}

async function withTimeout<T>(promise: Promise<T>, timeoutMs: number): Promise<T> {
  return await Promise.race([
    promise,
    new Promise<T>((_resolve, reject) => {
      setTimeout(() => reject(new Error(`timed out after ${timeoutMs}ms`)), timeoutMs);
    }),
  ]);
}
